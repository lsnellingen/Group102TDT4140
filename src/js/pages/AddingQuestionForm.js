// Libs
import React, { Component } from 'react';

class AddingQuestionForm extends Component {

  constructor(props) {
    super(props);
      this.state = {
        question: '',
        type: 'chooseType',
        currentAlternative: '',
        alternatives: [],
        showWarning: false
      };
    }

    handleFiltering(event) {
       if(event.target.value == "Yes or No") {
         this.setState({
           type: event.target.value,
           alternatives: ["Yes", "No"],
           showWarning: false
          })
       } else {
         this.setState({
           type: event.target.value,
           showWarning: false
          })
       }
    }

    handleQuestionChange(event) {
      this.setState({
        question: event.target.value,
        showWarning: false
       })
    }

    handleAlternativeChange(event) {
      this.setState({
        currentAlternative: event.target.value,
        showWarning: false
      })
    }

    handleAddAlternative(event) {
      event.preventDefault();
      var newAlternatives = this.state.alternatives;
      newAlternatives.push(this.state.currentAlternative);
      this.setState({
        alternatives: newAlternatives,
        currentAlternative: ""
      })
    }

    handleSubmit(event){
      event.preventDefault();
      if(this.state.question == '' || this.state.type == 'chooseType') {
        this.setState({showWarning: true})
      } else {
        if(this.state.type == 'Multiple choice' && this.state.alternatives.length == 0) {
          this.setState({showWarning: true})
        } else {
          const newQuestion = {"question": this.state.question, "type": this.state.type, "alternatives": this.state.alternatives}
          this.props.addQuestion(newQuestion);
          this.setState({
            question: '',
            type: '',
            alternatives: [],
            currentAlternative: ''
          })
        }
      }
    }

  render() {
    return (
      <div className="panel panel-info">
        <div className="panel-heading">
        Add new question:
        </div>
        <div className="panel-body">
          <div className="container">
            <div className="row">
              <div className="col-xs-7">
                <form onSubmit={this.handleSubmit.bind(this)}>

                  <div className="form-group">
                    <p>Question:</p>
                    <input type="text" className="form-control" id="question" placeholder="Question" value={this.state.question} onChange={this.handleQuestionChange.bind(this)}/>
                  </div>
                  <div className="form-group">
                    <p>Type of question:</p>
                    <select type="subject-selector" className="form-control" id="subject-selector" name="subject-selector" value={this.state.type} onChange={this.handleFiltering.bind(this)}>
                      <option value="chooseType">Choose type</option>
                      <option value="Text">Text</option>
                      <option value="Multiple choice">Multiple choice</option>
                      <option value="Yes or No">Yes or No</option>
                      <option value="Rating">Rating</option>
                    </select>
                  </div>
                  { this.state.type == 'Multiple choice' ?
                    <div className="form-group">
                    <hr />
                      <div className="row">
                        <div className="col-xs-8">
                          <p>Alternatives:</p>
                          <div className="form-inline">
                            <input type="text" className="form-control" id="question" placeholder="Alternative" value={this.state.currentAlternative} onChange={this.handleAlternativeChange.bind(this)} />
                            <button className="btn btn-info marginLeft20" onClick={this.handleAddAlternative.bind(this)}>Add alternative</button>
                          </div>
                        </div>
                        <div className="col-xs-4 margin-top20">
                          <p>Your alternatives: </p>
                            { this.state.alternatives.map(alternative => {
                              return <li className="questionAlternatives" key={alternative}>{alternative}</li>;
                            })}
                        </div>
                      </div>
                      <hr />
                    </div>
                    : null}
                    {this.state.showWarning ?
                      <div className="form-group removePadding">
                        <p className="alert alert-danger userMessage">You have to fill in all sections.</p>
                      </div>
                      : null }
                  <div className="form-group">
                    <button type="submit" className="btn btn-info hCenter formButtonMargin">Add question</button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}


export default AddingQuestionForm;
