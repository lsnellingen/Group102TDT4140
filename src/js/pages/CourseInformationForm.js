// Libs
import React, { Component } from 'react';
import axios from 'axios';

class CourseInformationForm extends Component {

  constructor(props) {
    super(props);
      this.state = {
        myCourses: [],
        showWarning: false,
        showMessage: false,
        course: 'chooseCourse',
        name: '',
        description: ''
      };
    }
    static contextTypes = {
      user: React.PropTypes.object
    };

    componentDidMount() {
      const currentUser = this.context.user ? this.context.user.email.toLowerCase() : 'No user';
      axios.get('/getCourses/' + currentUser)
      .then(res => {
        if(res.data[0].courses == "") {
          this.setState({
            myCourses: [],
            showWarning: true});
        } else {
          this.setState({ myCourses: res.data[0].courses.split('+') });
        }
      });
    }

    handleOptionChange(field,changeEvent){
      var object = {};
      object[field] = changeEvent.target.value;

      this.setState(object);
      this.setState({showMessage: false})
    }

    handleFiltering(event) {
      this.setState({
        course: event.target.value,
        showMessage: false
       })
    }

    handleSubmit(event){
      event.preventDefault();
      if(this.state.name == '' || this.state.description == '' || this.state.course == 'chooseCourse') {
        this.setState({showMessage: true})
      } else {
        const newQuery = {"name": this.state.name, "description": this.state.description, "course": this.state.course}
        this.props.updateQueryInformation(newQuery);
      }
    }

  render() {
    return (
      <div className="">
        <div className="panel-body">
          <div className="container">
            <div className="row">
              <div className="col-xs-8 col-xs-offset-1">
                <form onSubmit={this.handleSubmit.bind(this)}>
                  <div className="form-group">
                    <p>Choose course:</p>
                    <select type="subject-selector" className="form-control" id="subject-selector" name="subject-selector" value={this.state.course} onChange={this.handleFiltering.bind(this)}>
                      <option value="chooseCourse">Choose course</option>
                        { this.state.myCourses.map(course => {
                          return <option key={course} value={course} >{course}</option>;
                        })}
                    </select>
                    {this.state.showWarning ?
                      <div className="form-group removePadding">
                        <p className="alert alert-danger userMessage">You are not registered in any courses. Go to the course page and register for courses.</p>
                      </div>
                      : null }
                  </div>
                  <div className="form-group">
                    <p>Name:</p>
                    <input type="text" className="form-control" id="name" placeholder="Name" value={this.state.name} onChange={this.handleOptionChange.bind(this,"name")}/>
                  </div>
                  <div className="form-group">
                    <p>Short description: </p>
                    <textarea type="text" className="form-control" id="description" placeholder="Description" value={this.state.description} onChange={this.handleOptionChange.bind(this,"description")}/>
                  </div>
                  {this.state.showMessage ?
                    <div className="form-group removePadding">
                      <p className="alert alert-danger userMessage">You have to fill in all sections.</p>
                    </div>
                    : null }
                  <button type="submit" className="btn btn-info hCenter formButtonMargin mediumButton">Start adding questions</button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <hr />
      </div>
    );
  }
}




export default CourseInformationForm;
