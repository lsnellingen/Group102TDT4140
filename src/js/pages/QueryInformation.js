// Libs
import React, { Component } from 'react';

class QueryInformation extends Component {

  static contextTypes = {
    user: React.PropTypes.object
  };


  render() {
    return (
      <div className="panel panel-info">
        <div className="panel-heading">
          <h3 className="panel-title">Query information:</h3>
        </div>
        <div className="panel-body">
          <table className="table">
            <tbody>
              <tr>
                <th>Name:</th>
                <td>{this.props.query.name}</td>
              </tr>
              <tr>
                <th>Description:</th>
                <td>{this.props.query.description}</td>
              </tr>
              <tr>
                <th>Course:</th>
                <td>{this.props.query.course}</td>
              </tr>
              <tr>
                <th>Creator:</th>
                <td>{this.context.user ? this.context.user.email : null}</td>
              </tr>
              <tr>
                <th>Number of questions:</th>
                <td>{this.props.numberOfQuestions}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

export default QueryInformation;
