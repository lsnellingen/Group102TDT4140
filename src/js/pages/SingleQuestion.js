// Libs
import React, { Component } from 'react';

class SingleQuestion extends Component {


  render() {
    return (
      <div className="panel panel-info marginButtom10">
        <div className="panel-heading padding4">

        </div>
        <div className="panel-body removePadding">
        <div className="container-fluid">
        <table className="table fixedTable removeMargin">
          <tbody>
            <tr>
              <th className="col-xs-1">#</th>
              <th className="col-xs-6">Question:</th>
              <th className="col-xs-2">Type:</th>
              <th className="col-xs-3">Alternatives:</th>
            </tr>
            <tr>
              <td className="col-xs-1">{this.props.number}</td>
              <td className="col-xs-6">{this.props.question.question}</td>
              <td className="col-xs-2">{this.props.question.type}</td>
              <td className="col-xs-3">
                { this.props.question.alternatives.length > 0 ? this.props.question.alternatives.map(alternative => {
                  return <p className="removeMargin" key={alternative}>{alternative}</p>;
                }) : 'None'}
              </td>
            </tr>
          </tbody>
        </table>
        </div>
        </div>
      </div>
    );
  }
}




export default SingleQuestion;
