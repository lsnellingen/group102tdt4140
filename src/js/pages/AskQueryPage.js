import React, { Component } from 'react';
import DocumentTitle from 'react-document-title';
import axios from 'axios';

import SingleQuestion from './SingleQuestion';
import CourseInformationForm from './CourseInformationForm';
import AddingQuestionForm from './AddingQuestionForm';
import QueryInformation from './QueryInformation';

class AskQueryPage extends Component {

  constructor(props) {
    super(props);
      this.state = {
        status: 'addingQueryInformation', //addingQueryInformation, addingQuestions
        myCourses: [],
        currentCourse: '',
        showWarning: false,
        query: {"description": "I would like to know how many of you that are willing to wake up earlier to get more out of the day.", "name": "Earlier lecturers on friday", "course": "TDT4120 - Algoritmer og datastrukturer"},
        questions: []
      };
    }
    static contextTypes = {
      user: React.PropTypes.object
    };

    updateQueryInformation(newQuery) {
      this.setState({
        query: newQuery,
        status: 'addingQuestions'
      })
    }

    addQuestion(newQuestion) {
      var newQuestions = this.state.questions;
      newQuestions.push(newQuestion);
      this.setState({questions: newQuestions});
    }

    deleteQuestion(key) {
      var newList = this.state.questions;
      newList.splice(key, 1);
      this.setState({
        questions: newList
      })
    }

    handleQueryRequest(event) {
      event.preventDefault();
      const username = this.context.user.email;
      const name = this.state.query.name;
      const description = this.state.query.description;
      const course = this.state.query.course;
      axios.post('/sendQuery/' + username +
                 '/' + name +
                 '/' + description +
                 '/' + course)
      .then(res => {
        var id = res.data.insertId;
        console.log("Query done with ID: " + id);
        this.sendQueries(id);
      });
      this.setState({status: 'querySendt'})
    }

    sendQueries(id) {
      const questions = this.state.questions;
      for (let question of questions) {
        question.question = question.question.replace(/\?/g, 'QUESTIONMARK');
        if(question.alternatives.length > 0) {
          axios.post('/sendQueries/' + id +
                     '/' + question.question +
                     '/' + question.type +
                     '/' + question.alternatives.join('+'))
          .then(res => {
            console.log("Querie done with ID: " + res.data.insertId);
          });
        } else {
          axios.post('/sendQueries/' + id +
                     '/' + question.question +
                     '/' + question.type +
                     '/' + 'Empty')
          .then(res => {
            console.log("Querie done with ID: " + res.data.insertId);
          });
        }
      }
    }

  render() {
  	return (
  		<DocumentTitle title={`Ask query`}>
        <div className="container">
          <div className="row">

            <div className="col-xs-12">
              <br />
              <h3>Ask query</h3>
              <hr />
            </div>

            <div className="col-xs-11">
              { this.state.status == 'addingQueryInformation' ? <CourseInformationForm updateQueryInformation={this.updateQueryInformation.bind(this)}/> : null}
            </div>
            <div className="col-xs-8">
              { this.state.status == 'addingQuestions' ? <AddingQuestionForm addQuestion={this.addQuestion.bind(this)} /> : null }
              { this.state.status == 'querySendt' ? <p className="alert alert-success">Query successfully created. Information is still visible untill you leave this page. </p> : null}
              { this.state.questions.map(question => {
                return <SingleQuestion key={question.question} number={this.state.questions.indexOf(question) + 1} deleteQuestion={this.deleteQuestion.bind(this)} question={question} />;
              })}
            </div>
            <div className="col-xs-4">
              { this.state.status == 'addingQuestions' || this.state.status == 'querySendt' ?
                  <QueryInformation query={this.state.query} numberOfQuestions={this.state.questions.length} /> : null
               }
              { this.state.questions.length > 0 && this.state.status == 'addingQuestions' ?
                  <button className="btn btn-info bigButton" onClick={this.handleQueryRequest.bind(this)}>CREATE QUERY</button> : null
               }
            </div>
          </div>
        </div>
	    </DocumentTitle>
  	);
  }
}

export default AskQueryPage;
