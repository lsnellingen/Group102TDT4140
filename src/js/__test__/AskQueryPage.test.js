import React from 'react';
import ReactDOM from 'react-dom';
import AskQueryPage from '../pages/AskQueryPage';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

import SingleQuestion from '../pages/SingleQuestion';
import CourseInformationForm from '../pages/CourseInformationForm';
import AddingQuestionForm from '../pages/AddingQuestionForm';
import QueryInformation from '../pages/QueryInformation';

describe('AskQueryPage', () => {
  var ReactTestUtils = require('react-addons-test-utils') // ES5 with npm

    test('renders without crashing', () => {
      /*const renderer = ReactTestUtils.createRenderer();*/

      const tree = renderer.create(
        <AskQueryPage/>).toJSON();


      expect(tree).toMatchSnapshot();




    });

});
