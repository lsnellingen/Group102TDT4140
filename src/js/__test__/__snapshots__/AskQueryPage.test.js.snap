// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AskQueryPage renders without crashing 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <br />
      <h3>
        Ask query:
      </h3>
      <hr />
    </div>
    <div
      className="col-xs-11"
    >
      <div
        className=""
      >
        <div
          className="panel-body"
        >
          <div
            className="container"
          >
            <div
              className="row"
            >
              <div
                className="col-xs-8 col-xs-offset-1"
              >
                <form
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <p>
                      Choose course:
                    </p>
                    <select
                      className="form-control"
                      id="subject-selector"
                      name="subject-selector"
                      onChange={[Function]}
                      type="subject-selector"
                      value="chooseCourse"
                    >
                      <option
                        value="chooseCourse"
                      >
                        Choose course
                      </option>
                    </select>
                  </div>
                  <div
                    className="form-group"
                  >
                    <p>
                      Name:
                    </p>
                    <input
                      className="form-control"
                      id="name"
                      onChange={[Function]}
                      placeholder="Name"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <p>
                      Short description: 
                    </p>
                    <textarea
                      className="form-control"
                      id="description"
                      onChange={[Function]}
                      placeholder="Description"
                      type="text"
                      value=""
                    />
                  </div>
                  <button
                    className="btn btn-info hCenter formButtonMargin mediumButton"
                    type="submit"
                  >
                    Start adding questions
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <hr />
      </div>
    </div>
    <div
      className="col-xs-8"
    />
    <div
      className="col-xs-4"
    />
  </div>
</div>
`;
